File: Data Structures & Algorithms/search-2d-matrix/submission-6.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int[] nums = new int[matrix.length * matrix[0].length];
        int index = 0;
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                nums[index++] = matrix[i][j];
            }
        }
        return binarySearch(0, nums.length - 1, nums, target);
    }

    private boolean binarySearch(int l, int r, int[] nums, int target) {
        if (l > r) return false;
        int index = l + (r - l) / 2;
        if (nums[index] == target) return true;
        else if (nums[index] < target) return binarySearch(index + 1, r, nums, target);
        else {
            return binarySearch(l, index - 1, nums, target);
        }
    }
}

File: Data Structures & Algorithms/top-k-elements-in-list/submission-6.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int, int> freqs;
        vector<vector<int>> buckets(nums.size() + 1);
        for (int n : nums) {
            freqs[n]++;
        }
        for (const auto& pair : freqs) {
            buckets[pair.second].push_back(pair.first);
        }
        vector<int> result(k);
        int index = 0;
        for (int i = buckets.size() - 1; i >= 0; i--) {
            for (int n : buckets[i]) {
                result[index] = n;
                index++;
                if (index == k) return result;
            }
        }
        return result;
    }
};
